### PR TITLE
[FIRRTL] GCT View Scattering in LowerAnnotations

### PR DESCRIPTION
**Pull request by seldridge**
Move all logic related to Grand Central (GCT) view annotation scattering
from FIRAnnotations into the LowerAnnotations pass.  In this commit
there are difficult to spot changes to parseAugmentedType and
applyGCTView functions (where the latter is extracted from the body of a
method inside FIRAnnotations):
    
1. Functions now use ApplyState instead of parameters passed to the
   functions.
    
2. Any scattered annotations are handled by using ApplyState's ability
   to add annotations to the worklist.  E.g., parseAugmentedType will add a
   new annotation to the worklist for every ground type leaf.

Change firtool to always run LowerAnnotations.  Remove the "--new-anno"
option to conditionally run LowerAnnotations.
    
Signed-off-by: Schuyler Eldridge <schuyler.eldridge@sifive.com>

### CODE REVIEW

**Comment by seldridge** [comment]
This needs updated tests. Any annotations.fir/annotations-errors.fir tests need to be moved into tests for the lower annotations pass (as these are now unprocessed and just moved to a `rawAnnotations` attribute on the circuit). 

This also likely has a bug as the SFC integration test `Wire.fir` is not currently passing (and it should for this to land).

The prerequisite work for this pass landed in:
- bcff7bcc
- 02e9c015
- d5293a95
- 3efb275a
- 462a7e77

The commit unconditionally enabling `LowerAnnotations` should also be a no-op that can land separately if needed.

**Comment by seldridge** [comment]
The bug is likely in the fact that this isn't applying the don't touches to the leaves. (We should preserve this in this PR.) Additionally, there is some different handling around targets which may be problematic. The old code has special handling for sub targets while this does not.

**Comment by seldridge** [comment]
There are a few prerequisite commits stacked in this patch. Specifically, `LowerAnnotations` uses `circt.fieldID` instead of `SubAnnotationAttr` to encode annotations which target part of an aggregate. This requires changes to downstream passes that look for `SubAnnotationAttr`: (1) LowerTypes to lower `circt.fieldID` in the same way as `SubAnnotationAttr` and (2) FlattenMemory to use `circt.fieldID` as an indication to skip processing as it does with `SubAnnotationAttr`.

There is also a bug fix related to the `resolveEntities` utility where it expects the leaf module to always be a module when this could be any module-like.

Tests are migrated roughly in-place from `annotations.fir` into `annotation.mlir`. Critically, this patch requires _no modifications_ to the existing GCT integration tests.

**Comment by seldridge** [comment]
(Mentioned offline:) Thanks for fixing the crash @dtzSiFive. 💯  There was one other bug exposed on an internal design where an NLA don't touch on a subfield/subindex was erroring out because this patch indicated that that case was an illegal NLA. That is fixed in the force push. I'm seeing one test of a large internal design pass so I have some confidence that this is not actually breaking anything.

File: lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp [hunk 1]
```diff
@@ -267,7 +293,15 @@ static const llvm::StringMap<AnnoRecord> annotationRecords{{
...
-    {"circt.missing", {tryResolve, applyWithoutTarget<>}}
+    {"circt.missing", {tryResolve, applyWithoutTarget<>}},
+    // Grand Central Views/Interfaces Annotations
+    {serializedViewAnnoClass, {noResolve, applyGCTView}},
+    {viewAnnoClass, {noResolve, applyGCTView}},
+    {companionAnnoClass, {stdResolve, applyWithoutTarget<>}},
+    {parentAnnoClass, {stdResolve, applyWithoutTarget<>}},
+    {augmentedGroundTypeClass, {stdResolve, applyWithoutTarget<true>}},
+    // Miscellaneous Annotations
+    {dontTouchAnnoClass, {stdResolve, applyDontTouch}}
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp:304]: I love how this allows us to factor out the scattering logic and move it closer to the passes that actually make use of it.

File: test/Dialect/FIRRTL/annotations.fir [hunk 2]
```diff
@@ -708,316 +657,6 @@ circuit GCTMemTap : %[
 
 ; // -----
 
-; Test sifive.enterprise.grandcentral.ViewAnnotation
-circuit GCTInterface : %[
-[
-  {
-    "class": "sifive.enterprise.grandcentral.GrandCentralView$SerializedViewAnnotation",
-    "name": "view",
```
Comment (fabianschuiki) [inline test/Dialect/FIRRTL/annotations.fir:716]: This is now covered in [test/Dialect/FIRRTL/annotations.mlir](https://github.com/llvm/circt/pull/3141/files#diff-c3ad6ada6386fbfc94203d02aec933bfe504a376b9aaa5392d78325eac978b40R188-R226), right? Just to make sure we don't lose the test coverage.

**Review by fabianschuiki — APPROVED** [review]
LGTM! Breaking up the big chunk of annotation scattering code is great. Looking forward to applying this to GCT Data Taps as well :+1:

File: test/Dialect/FIRRTL/annotations.fir [hunk 3]
```diff
@@ -380,57 +380,6 @@ circuit Foo: %[[{"target": "~Foo|Foo>_T_0", "a": "a"},
...
-      reset => (reset, UInt<4>("h0"))
-    ; CHECK: %_T_4 = chirrtl.seqmem sym @_T_4 Undefined
-    smem _T_4 : UInt<1>[9] [256]
-    ; CHECK: %_T_5 = chirrtl.combmem sym @_T_5
-    cmem _T_5 : UInt<1>[9] [256]
-    ; CHECK: chirrtl.memoryport Infer %_T_5 {annotations =
-    ; CHECK-SAME: {class = "firrtl.transforms.DontTouchAnnotation"}
-    infer mport _T_6 = _T_5[reset], clock
-    ; CHECK: firrtl.instance _T_7 sym @_T_7
-    inst _T_7 of Bar
```
Comment (seldridge) [inline test/Dialect/FIRRTL/annotations.fir:417]: I dropped this part of this test as SFC will not let you put a `DontTouchAnnotation` on a module or instance.

**Pull request merged**